CONTRIBUTING.md

```diff
@@ -6,23 +6,119 @@ JupyterHub also follows the Jupyter [Community Guides](https://jupyter.readthedo
+or a similar system to run [Ubuntu Linux](https://www.ubuntu.com/)
+for development.
+
+### Install Python
+
+JupyterHub is written in the [Python](https://python3.org) programming
+language, and requires you have at least version 3.5 installed locally.
+If you haven't installed Python before, the recommended way to install
+it is to use [miniconda](https://conda.io/miniconda.html). Remember
+to get the 'Python 3' version, and **not** the 'Python 2' version!
+
+### Install nodejs
+
+`configurable-http-proxy`, the default proxy implementation for JupyterHub,
+is written in Javascript to run on [NodeJS](https://nodejs.org/en/). If
+you have not installed nodejs before, we recommend installing it in the
+`miniconda` environment you set up for Python. You can do so with
+`conda install nodejs`.
+
+### Install git
+
+JupyterHub uses [git](https://git-scm.com) & [GitHub](https://github.com)
+for development & collaboration. You need to [install git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
+to work on JupyterHub. We also recommend getting a free account
+on GitHub.com.
+
+## Set up a development install
+
+When developing JupyterHub, you need to make changes to the code
+& see their effects quickly. You need to do a developer install
+to make that happen.
+
+1. Clone the [JupyterHub git repository](https://github.com/jupyterhub/jupyterhub)
+   on to your computer.
+
+   ```bash
+   git clone https://github.com/jupyterhub/jupyterhub
+   cd jupyterhub
+   ```
+
+2. Make sure the `python` you installed and the `npm` you installed
+   are available to you on the commandline.
+
+   ```bash
+   python -V
+   ```
+
+   This should return a version number greater than or equal to 3.5.
+
+   ```bash
+   npm -v
+   ```
+
+   This should return a version number greater than or equal to 5.0.
 
-### Troubleshooting a development install
+3. Install `configurable-http-proxy`. This is required to run JupyterHub.
 
-If the `pip3 install` command fails and complains about `lessc` being
-unavailable, you may need to explicitly install some additional JavaScript
-dependencies:
+   ```bash
+   npm install -g configurable-http-proxy
+   ```
 
-    npm install
+   If you get an error that says `Error: EACCES: permission denied`,
+   you might need to prefix the command with `sudo`. If you do not have
+   access to sudo, you may instead run the following commands:
+
+   ```bash
+   npm install configurable-http-proxy
+   export PATH=$PATH:$(pwd)/node_modules/.bin
+   ```
+
+   The second line needs to be run every time you open a new terminal.
+
+4. Install the python packages required for JupyterHub development.
+
+   ```bash
+   pip3 install -r dev-requirements.txt
+   pip3 install -r requirements.txt
+   ```
+
+5. Install the development version of JupyterHub. This lets you edit
+   JupyterHub code in a text editor & restart the JupyterHub process
+   to see your code changes immediately.
+
+   ```bash
+   pip3 install --editable .
```

Comment:
Thanks for writing out `--editable` rather than `-e`! Especially in this context.